monashspa/PHS2061/PHS2061_Q3_Synthesiser.py

```diff
@@ -0,0 +1,165 @@
+# Q3: Fourier analysis
+# Part 2: Synthesising an instrument
+# ------------------------------------------------------------------------------
+
+import numpy as np
+import matplotlib.pyplot as plt 
+import scipy.io.wavfile
+# from IPython.display import Audio
+from matplotlib.widgets import Cursor
+
+# ------------------------------------------------------------------------------
+# We're going to create a set of sin functions. 
+# You can specify the amplitudes, frequencies and phases shifts for each of the waves. 
+# Try adding more of the harmonics you found when analysing your recordings.
+# ------------------------------------------------------------------------------
+
+
+# Samples per second
+sample_rate = 44100
+
+# Duration in seconds
+duration_s = 5
+
+## Create a list of sine waves to add together 
+## Each one has an ampltude, frequency (Hz) and phase (radians)
+wave_properties = [
+    (0.4, 100.0, 0.0),
+    (0.1, 200.0, np.pi/2.0),
+    (0.005, 400.0, 0.0),
+]
+
+## Create the waveforms for each sine wave
+waveforms = []
+sample_number = np.arange(duration_s * sample_rate)
+for amp,freq,phase in wave_properties:
+    waveforms += [amp*np.sin(2 * np.pi * sample_number * freq / sample_rate + phase)]
+
+# ------------------------------------------------------------------------------
+# Now we can plot each of the waves we've created, along with the total sum.
+# ------------------------------------------------------------------------------
+
+## Plot a restricted range of the samples
+first_sample = 0
+last_sample  = 1000
+
+total = np.zeros(duration_s * sample_rate)
+
+fig, (ax0, ax1) = plt.subplots(1, 2)
+for wave in waveforms:
+    total += wave
+    x = sample_number[first_sample:last_sample]
+    y = wave[first_sample:last_sample]
+    ax0.plot(x,y)
+ax0.set_xlabel('Sample number')
+ax0.set_ylabel('Amplitude')
+ax0.set_title('Individual waves')
+
+
+x = sample_number[first_sample:last_sample]
+y = total[first_sample:last_sample]
+ax1.plot(x,y)
+ax1.set_xlabel('Sample number')
+ax1.set_title('Total wave')
+
+fig.suptitle('Synthesised waveforms')
+
+# save the plot
+plt.savefig('part2_sine_waves.pdf')
+
+## Save the sample and see what it sounds like!
+scipy.io.wavfile.write("waveform_unwindowed.wav", sample_rate, total)
+
+# ------------------------------------------------------------------------------
+# To make the wave sound more like the guitar string, we can multiple it be an
+# envelope function to mimic how it moves when we pluck the string, and as the 
+# energy is disipated over time. 
+# ------------------------------------------------------------------------------
+
+## Lets use an exponential function to model the decay
+## and 1-exp to model the pluck
+
+decay_time = 1.0 ## seconds 
+decay_shape = np.exp(-sample_number/(decay_time*sample_rate) )
+
+attack_time = 0.01 ## seconds 
+attack_shape = 1-np.exp(-sample_number/(attack_time*sample_rate) )
+
+windowed = np.multiply(total,np.multiply(decay_shape,attack_shape))
```

Comment:
Can you not just write this as `windowed = total * attack_time * decay_time`

Reply:
Yes that is much simpler, fixed